hydromt/components/kernel_config.py

```diff
@@ -0,0 +1,177 @@
+"""A component to write configuration files for simulations/kernels."""
+
+from os import makedirs
+from os.path import abspath, dirname, isabs, join, splitext
+from pathlib import Path
+from typing import TYPE_CHECKING, Any, Dict, Optional
+
+from hydromt.components.base import ModelComponent
+from hydromt.io.path import make_config_paths_relative
+from hydromt.io.readers import read_yaml
+from hydromt.io.writers import write_toml, write_yaml
+
+if TYPE_CHECKING:
+    from hydromt.models import Model
+
+DEFAULT_KERNEL_CONFIG_PATH = "kernel_config.yml"
+
+
+class KernelConfigComponent(ModelComponent):
+    """A component to write configuration files for simulations/kernels."""
+
+    def __init__(
+        self,
+        model: "Model",
+    ):
+        """Initialize a KernelConfigComponent.
+
+        Parameters
+        ----------
+        model: Model
+            HydroMT model instance
+        """
+        self._data: Optional[Dict[str, Any]] = None
+        super().__init__(model=model)
+
+    @property
+    def data(self) -> Dict[str, Any]:
+        """Kernel config values."""
+        if self._data is None:
+            self._initialize_kernel_config()
+        if self._data is None:
+            raise RuntimeError("Could not load data for kernel config component")
+        else:
+            return self._data
+
+    def _initialize_kernel_config(self, skip_read=False) -> None:
+        """Initialize the kernel configs."""
+        if self._data is None:
+            self._data = dict()
+            if self._root.is_reading_mode() and not skip_read:
+                self.read()
```

Comment:
ah I see, I understand better now. I'm honestly happy with whatever you think is best. At the very least I'll add the functionality to read from a template as that will be necessary anyway.

Reply:
I've added the reading regardless of mode and written some tests for it. From these comments I also deduced that a model should have a kernel config component by default so I added it during init, please let me know if I have misunderstood. I think this also is a bit closer now to what you wanted re it being overridable for plugins since they can now just change the default string that gets added in the model init either by subclassing the component or by just providing a different string in their model init. Is this what closer to what you wanted?

Reply:
I would like to have Dirk advice on how to deal with config before merging.
I still don't like the ``DEFAULT_KERNEL_CONFIG_PATH``. I think most models actually don't need to subclass ``ConfigComponent`` as functionality is pretty standard and we support reading and writing from TOML. But if I'm wflow, then I still need to subclass to only change the ``DEFAULT_KERNEL_CONFIG_PATH``.... That users can overwrite the template is good but plugins should be able to overwrite the real default value as it does not make sense for them to use the hydromt one.

Reply:
This is true for the code that this comment is attached to, that I understand. However, in the newest version I changed this, so that the component actually indeed stores the path that can be given at model init, meaning that all a plugin has to do is changed what they pass into the config component. Hopefully this is better, but I'm okay with waiting for what @DirkEilander thinks.

Reply:
Regarding `DEFAULT_KERNEL_CONFIG_PATH`. From the code I don't understand how the functionality you describes actually works @savente93. Could you give a short code example of how this would look like for a plugin? We used to this this with class properties as mentioned by @hboisgon. I don't really care how, but it would be good if plugins don't need to subclass to only change one class property.

Reply:
> So even in write/build mode, we would read from a template config file generally stored in the plugin repositories. Hence even in write mode, there is always a config to read so we did not add a check on self._root.is_reading_mode() only on skip_read. This is also why the config_fn filename was the only filename argument in Model.__init__.

> We could decide to make this behavior more transparent by adding a create method to ConfigComponent which creates a config from a template. Later this config can be updated by the not yet transfered setup_config method. It's a little bit of a breaking change for users but I think in a way better for transparency. @DirkEilander what do you think?

 I agree with @hboisgon that a `create` method makes sense and is consistent with other model components.